src/graph/type_indexededgelist.c

```diff
@@ -783,61 +807,126 @@ int igraph_neighbors(const igraph_t *graph, igraph_vector_t *neis, igraph_intege
-            eid2 = (long int) VECTOR(graph->ii)[i2];
-            n1 = (long int) VECTOR(graph->to)[eid1];
-            n2 = (long int) VECTOR(graph->from)[eid2];
-            if (n1 < n2) {
-                VECTOR(*neis)[idx++] = n1;
-                i1++;
-            } else if (n1 > n2) {
-                VECTOR(*neis)[idx++] = n2;
-                i2++;
-            } else {
-                VECTOR(*neis)[idx++] = n1;
-                VECTOR(*neis)[idx++] = n2;
-                i1++;
-                i2++;
+
+        if (multiple == IGRAPH_MULTIPLE && loops == IGRAPH_LOOPS_TWICE) {
+            while (i1 < j1 && i2 < j2) {
+                eid1 = (long int) VECTOR(graph->oi)[i1];
+                eid2 = (long int) VECTOR(graph->ii)[i2];
+                n1 = (long int) VECTOR(graph->to)[eid1];
+                n2 = (long int) VECTOR(graph->from)[eid2];
+                if (n1 < n2) {
+                    VECTOR(*neis)[idx++] = n1;
+                    i1++;
+                } else if (n1 > n2) {
+                    VECTOR(*neis)[idx++] = n2;
+                    i2++;
+                } else {
+                    VECTOR(*neis)[idx++] = n1;
+                    VECTOR(*neis)[idx++] = n2;
+                    i1++;
+                    i2++;
+                }
             }
-        }
 
-        while (i1 < j1) {
-            eid1 = (long int) VECTOR(graph->oi)[i1++];
-            VECTOR(*neis)[idx++] = (long int) VECTOR(graph->to)[eid1];
-        }
+            while (i1 < j1) {
+                eid1 = (long int) VECTOR(graph->oi)[i1++];
+                VECTOR(*neis)[idx++] = (long int) VECTOR(graph->to)[eid1];
+            }
+
+            while (i2 < j2) {
+                eid2 = (long int) VECTOR(graph->ii)[i2++];
+                VECTOR(*neis)[idx++] = (long int) VECTOR(graph->from)[eid2];
+            }
+        } else {
+            while (i1 < j1 && i2 < j2) {
+                eid1 = (long int) VECTOR(graph->oi)[i1];
+                eid2 = (long int) VECTOR(graph->ii)[i2];
+                n1 = (long int) VECTOR(graph->to)[eid1];
+                n2 = (long int) VECTOR(graph->from)[eid2];
+                if (n1 < n2) {
+                    i1++;
+                    IGRAPH_I_NEIGHBORS_SKIP(n1, 1);
+                    VECTOR(*neis)[idx++] = n1;
+                } else if (n1 > n2) {
+                    i2++;
+                    IGRAPH_I_NEIGHBORS_SKIP(n2, 1);
+                    VECTOR(*neis)[idx++] = n2;
+                } else {
+                    i1++;
+                    i2++;
+                    IGRAPH_I_NEIGHBORS_SKIP(n1, 2);
+                    VECTOR(*neis)[idx++] = n1;
+                    if ((loops == IGRAPH_LOOPS_ONCE && n1 == pnode && last_added == pnode) ||
+                         (multiple == IGRAPH_NO_MULTIPLE)) {
+                         length--;
+                         continue;
+                    }
+                    VECTOR(*neis)[idx++] = n2;
```

Comment:
Not sure if you see this; Codecov indicates that this branch was not covered by tests. It would be great if you could add a test case to cover this  branch.

Reply:
Yes, I see them :slightly_smiling_face: A great feature, and I'll try to fix them.